### DISCUSSION TITLE
[Design Proposal] Alert and Report Modal Redesign + Filter and Tab Configuration

### DISCUSSION THREAD

**Discussion by yousoph** [apache/superset#25729] [Ideas]
## Context
There have been a few feature requests around alerts and reports around providing more filter and/or tab details when setting up a report or alert. 

See the below for details on the feature requests: 
Filters and tabs: https://github.com/apache/superset/discussions/18472 
Tabs: https://github.com/apache/superset/discussions/18475 
Permalink to provide additional filter and tab info: [https://github.com/apache/superset/discussions/24843 and](https://github.com/apache/superset/discussions/24843) https://github.com/apache/superset/discussions/22905 
Send test report feature: https://github.com/apache/superset/discussions/18472#discussioncomment-2097851

We wanted to propose some UI updates to the alert and report creation flow that will include the option to set up filters and tabs, as well as update the grouping of the fields to improve the clarity of the modal and streamline the configuration of the reports/alerts. 

## Current Design
Alerts: 
![alerts_current](https://github.com/apache/superset/assets/10627051/d02a3cd8-c0ed-4665-9c0d-045452f09727)

Reports: 
![reports_current](https://github.com/apache/superset/assets/10627051/3873190c-0ba7-4f2c-99de-c41e94d5cb2a)

## Goals

1. **Decluttering the Interface:** The current modal appears cluttered, which can be overwhelming for users. This proposal aims to provide a cleaner, more organized layout.
2. **Linear Setup Process:** Present the user with the necessary options in a step-by-step manner to make alert/report set up more intuitive. 
3. **New Features:** 
    1. **Predefined Filter/Tab Selection:** Introduce the ability to specify a predefined filter or tab for the alert/report 
    2. **Send a Test report**: Introduce the ability to send a report for testing, outside of the report schedule

## Proposed Changes

![Add report](https://github.com/apache/superset/assets/10627051/851b0de4-1b06-4e29-87d0-b448604228ea)
![report_contents](https://github.com/apache/superset/assets/10627051/ad8f02ca-da80-4d7e-92f3-0e5b6f9903b0)
![report_schedule](https://github.com/apache/superset/assets/10627051/6e13b396-475d-4253-81b3-240667340011)
![image](https://github.com/apache/superset/assets/10627051/1068eb65-b0bf-4d3d-ac6c-67faa81477bb)

### 1. Fields grouped into sections
The fields in the modal will be grouped into logical categories, similar to the pattern used in database connection modal. This categorization will improve the user's ability to quickly locate and interact with the relevant options.

### 2. Step-by-Step Setup Process
Instead of displaying all options at once, the modal will show a step by step setup process, with each step clearly indicating the information required. Users can still jump around between sections if desired, but the modal focuses on one section at a time. 

### 3. New Feature - Filter Configuration
When a dashboard is selected, users will be able to configure filters and filter values for the report using the filters that are already defined on the dashboard. Multiple filters can be selected. 

### 4. New Feature - Tab Configuration
When a dashboard with multiple tabs is selected, the user will be able to specify the tab that should be used for the report. 

### 5. New feature - Send test alert/report
Users are able to send a test alert/report to themselves, all recipients, or owners of the report to verify their settings.

## Future Considerations
We felt the dropdown methods to set up filters and tabs would be more intuitive for users than permalinks would be. However, permalinks could still be a future addition to this modal to solve for anchored charts, other URL params, etc.

**Comment by rusackas** [comment]
This all seems great. The one thing I'd like to ask for in this proposal is that if we make these changes, please update/append the existing documentation. Otherwise, these all seem like welcome improvements!

**Comment by TechAuditBI** [comment]
Awsome changes! 
I'd like to propose a couple ideas on top: 
1. Show full modal upon creating a report from a chart edit mode. Our users usually create reports for multiple users, so it is very uncomfortable for them to create a report and then to go to reports list and add a couple e-mails into recipients field.
2. Send description or some new field as a plain text in a report. Usually our users recieve about 5 reports per hour so it is quite useful for them to get some info about what they've recieved.
3. Ability to send multiple charts/dashboards in a single report. Sometimes users have to send multiple charts comined by the same theme. Currently there is no other option then creating multiple reports which is very cmplicated to support.
4. Ability to send multiple content types at the same time. For example send a chart png and a csv with it's data in one report.

Also a bunch of thoughts on permalinks and overall usecases. 
As for now we've been using reports heavily. Currently we have 250+ scheduled reports. Most of the time our users prefer to personalize (f.e. filter data according to their department and remove a bunch of unneccesary columns/metrics) a couple of charts from a dashboard and make a report out of them. Before we made permalinks sendable our users were forced to create a copy of their charts after customising them. That is also why users can't combine their charts into dashboard. Which made reports completely impossible to support. Permalinks  have improved our life a bit but there is still a problem of transfering changes made in original chart into reports. As for now changes in the original chart won't transder into customised one in reports. We couldn't find a simple solution for that yet.
I'm not shure if our experience is actually suitable for everyone. So please give me some feedback on this. Maybe we just did something wrong :)

**Reply by yousoph** [reply]
Hi @TechAuditBI ! 

Thanks for the feedback! All these features sound useful and any contributions to add these features would be welcomed! I think the redesigned modal could be easily adjusted to support the features that you mentioned when they’re ready to be implemented. 

For #1, the modal from the chart/dashboard view are simplified to make it easy to set up a single report for the current user- perhaps a link the full view could be the way to go.

As for the permalink use case, would the dashboard filters on reports help in this case once the feature is implemented?

**Comment by puridach-w** [comment]
This all look great changes.
I might consider extending the scope of this SIP to encompass an additional feature: customizable email subject lines for report and alert notifications.

This feature allows users to customize the email subject of a report or alert. In the report submission creation interface, add a text area for entering the email subject. If a user inputs a value in this field, the email's subject will be this value. If not, the system will assign a default email subject line.

This is what it look like when apply to the current create report/alert modal:

<img width="1172" alt="Screenshot 2566-11-10 at 14 09 34" src="https://github.com/apache/superset/assets/76474644/4be9a5ba-9154-4401-98f7-689621a46b25">

### **This change need migration**
To incorporate this feature, we will need to initiate a database migration to add columns dedicated to storing email subject lines. 

More information: [customize email subject name](https://github.com/apache/superset/issues/25945)

**Reply by yousoph** [reply]
Hi @puridach-w , that would be a welcome contribution! I think this could fit nicely in the updated modal design as well in a similar location to what you have proposed, like this: 
![report mockup](https://github.com/apache/superset/assets/10627051/dc0228fa-1dc5-4260-9d5c-f504df8a31a9)

**Reply by puridach-w** [reply]
Hi @yousoph, thank for the design. So I can directly create PR for customize email subject name, right?

**Reply by yousoph** [reply]
Yes :) There should be some PRs up for the UI change proposed here soon, if you want to wait for that to be done before adding the email subject field

**Comment by zuzana-vej** [comment]
Hi @yousoph, thank you for proposing the above changes and improvements to the alert/report - they are really exciting and address a number of user feature requests. We have some feedback points we would like to share and are happy to discuss further to make these feature improvements bullet proof for the long term. 

**On the UI improvements (points 1&2):**
Very minor feedback: 

- Generally we feel laptop screens are wide and using a wider setup modal would allow users to see more settings in the same modal, rather than clicking through the narrow setup modal. The step by step process works well if users are going through it once to set it all up, but it can get more time consuming if they go back and forth. **We noticed the same modal pattern in the database configurations but they differ from the other modals in the application. Is there a specific reason why we are narrowing our modals?** 
- If the designers are set on the collapsible modal, one suggestion we have is to add indicators to see (1) which section is not filled and (2) if any section has error


**On “3. New Feature - Filter Configuration”:**
This is a very common feature request, but filters can get quite complex. We feel this needs some detailed thinking through the solution to ensure it works and scales.

Few questions we had:

- Are these filters the ones available in the dashboard, or would they be new filters?
- Some challenges we foresee: 
- - How do we handle a large number of filters? What if the dashboard has 15 or more filters? → do users need to specify all - do we display all filters, or do they just select a limited number of filters they want to apply? What if they want to apply more than one filter - would there be a limit to how many the UI can display? It seems like the setup process can be quite lengthy and complicated if dashboards have a large number of filters.
- - Filter scopes: Some filters are only applicable to some tabs, which can make the setup complex. Do we show only the filter applicable to the default tab, or how do we handle this alongside the tab selector feature? 
- - Filter that depends on another filter: some filters dynamically change based on another filter selection. How would this behavior reflect here?


There was a past proposal to use [permalinks](https://github.com/apache/superset/discussions/24843), which didn’t go through with the reasoning that it’s not too intuitive for users (users don’t have visibility of the dashboard state used to generate the permalink, and to make changes they need to create a new permalink). 

One potential solution we wanted to propose, which handles some of the above complexity, and gets around the permalink user experience challenges – What if we introduce a feature in the dashboard to set a report based on the Dashboard’s current state? Users would have a button in the dashboard that would internally generate the permalink. In the email alerting feature, users would be able to see a preview (screenshot) of the dashboard, and if they want to make changes to the email report, there's a button to take them back to the dashboard where they can edit and save the dashboard state (internally this means updating the permalink). Given that we're using the dashboard to generate the reports, we're able to handle really specific use cases such as filters with dependencies, global and nested tab selection, cross-filtering, etc.


**On “4. New Feature - Tab Configuration”:**

Similarly as with filters, tab configuration is a very common feature request, but it can get quite complex. We feel this needs some additional thinking to ensure it works - especially for nested tabs.

- Would there be an option to select **multiple nested tabs**? 

- - How do you enable users selecting nested tab(s) - there is currently no limit in dashboard to number of nested tabs that dashboard may include

-  - Do you enable users to select multiple tabs into the same report - creating multiple attachments? And is there a limit to how many they can select? 

**On “New feature - Send test alert/report”:**
Sounds good for a report.

How will this feature work for alert? What is the expectation users may want to see for alert? (if the threshold is not met - the alert wouldn’t fire → do users want to see how the report would look if the condition was met, OR do they want to be able to validate their SQL)? Both? 

Will the test email show in the activity log?

Thanks! 
@john-bodley , @justinpark , @michael-s-molina , @zuzana-vej

**Reply by yousoph** [reply]
Hi @zuzana-vej , thanks for all the feedback! 

* On modal width, the idea is to move towards a linear process for setting up the reports/alerts and less clutter on the screen. The current screens/modals have a lot of information that can get overwhelming. We think it will also be easier to scale now that there are different sections, and it shouldn't be too hard to jump between sections. 
* There is a green check mark (see first screenshot in the proposal) on sections that are completed, and we want to incorporate a hover on the disabled Create/Update button to show which fields are not yet complete: 
![image](https://github.com/apache/superset/assets/10627051/b12d0df7-74cd-47d2-b921-7cb0b050bf75)
* Filters: the list in the modal would be from existing dashboard filters (no new filter creation here at this point) 
* It's not included here, but we were also thinking that from the dashboard view, we could allow setting up the report based on the current tabs/filters that the user has applied there. We're thinking that the modal to set up the report would open with those filters/tabs as the selected values but users could adjust them if needed. 
* Tabs: yes - there would be some way to select nested tabs here. Either in a Tab 1 > Tab 2 type line item in the dropdown, or perhaps some nested view in the dropdown. We were thinking that users are currently only able to select on tab per report - though a future addition for multiple attachments could be useful. 
* Send now - I think we need to update the text in the mockup, but we were thinking about it for reports to start with. If we wanted to add it for alerts, I'd expect it just sends the report without checking the condition, but open to feedback of course.

**Comment by mayanklalwani** [comment]
Hi, I am curious about this feature. Can you please let me know if there already is some PR to incorporate this change or is this still under development?

**Reply by yousoph** [reply]
Hi @mayanklalwani 

A PR was just merged yesterday to add the tab selection: https://github.com/apache/superset/pull/29096 
Filters selection has yet to be developed

**Comment by bob-rescale** [comment]
Is there any rough timeline for when the Filter Configuration feature will be added? This would be very useful and I see it's mentioned in the [4.0 release notes](https://preset.io/blog/apache-superset-4-0-release-notes/#alerts-and-reports-modal-redesign) as coming in a future release. Thanks!

**Comment by neeravsalaria** [comment]
Any update on this feature - the filters selection ? When will it be released?

**Comment by villebro** [comment]
@yousoph is this feature still being worked on? If not, I would almost propose implementing the [permalink](https://github.com/apache/superset/discussions/24843) solution, as it's very easy to implement, and would unblock orgs that need this feature.
